### PR TITLE
Refactor caches to live in PredictionStrategy

### PR DESCRIPTION
**Pull request by jacobrgardner**
Step two towards handling fantasy observations.

- `mean_cache` and `covar_cache` are now properties of `PredictionStrategies`, and `ExactGP` caches a prediction strategy rather than the mean and covariance caches.
- Introduces a new `named_cached` decorator that allows us to specify string names for the cache storage, which lets us modify the cache later. Right now, this is used once in `InterpolatedLazyTensor` to delete the covar cache if the beta features (`fast_pred_var` / `fast_pred_samples` have changed). 
- `PredictionStrategy` objects are now created using all relevant test independent items, simplifying the interface of `exact_predictive_mean` and `exact_predictive_covar`.
- `train_train_covar` is now also cached, avoiding computing the train-train covariance matrix a second time.

Eventually, I would like to additionally move the batch mode and multi task reshaping out of `ExactGP.__call__` and in to the creation of the prediction strategy.

### CODE REVIEW

File: gpytorch/models/exact_gp.py [hunk 1]
```diff
@@ -78,30 +75,37 @@ def __call__(self, *args, **kwargs):
...
 
+        # Prior mode
+        elif self.train_inputs is None or self.train_targets is None:
+            full_inputs = args
+            full_output = super(ExactGP, self).__call__(*full_inputs, **kwargs)
+            if settings.debug().on():
+                if not isinstance(full_output, MultivariateNormal):
+                    raise RuntimeError("ExactGP.forward must return a MultivariateNormal")
+            full_mean, full_covar = full_output.mean, full_output.lazy_covariance_matrix
+            return full_output.__class__(full_mean, full_covar)
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply gpytorch/models/exact_gp.py]: Foolishness, mostly :-)

File: gpytorch/utils/memoize.py [hunk 2]
```diff
@@ -15,3 +15,19 @@ def g(self):
         return self.__cache[f]
 
     return g
+
+
+def named_cached(name):
```
[the remark this answers is not in the document]
Comment (jacobrgardner) [reply gpytorch/utils/memoize.py]: Turns out yes, `functools.partial` lets us accomplish this quite nicely!

File: gpytorch/models/exact_prediction_strategies.py [hunk 3]
```diff
@@ -17,14 +19,24 @@ def decorator(cls):
     return decorator
 
 
-def prediction_strategy(train_train_covar):
+def prediction_strategy(
+    num_train, train_inputs, train_mean, train_train_covar, train_labels, likelihood, non_batch_train=False
```
[the remark this answers is not in the document]
Comment (Balandat) [reply gpytorch/models/exact_prediction_strategies.py]: Ideally, we'd move to being consistent about every model having both a
- a `num_out: int` property specifying the number of outputs (1 in the basic case, the number of tasks in the multitask case)
- a `batch_shape: torch.Size` propoerty specifying the (potentially multi-dimensional) batch shape

Then if we internally always have a `batch_shape x d` representation of the features, and a `batch_shape x num_out` representation of the targets, we could get rid of a lot of the case distinctions by using `view` etc.

[the remark this answers is not in the document]
Comment (gpleiss) [reply gpytorch/models/exact_prediction_strategies.py]: I agree with both of those properties. `num_out` we could maybe do right now, `batch_shape` might be part of the larger refactor for getting arbitrary batch sizes (though we might be able to do something along those lines right now).

**Pull request merged**